agent/artifact_downloader.go

```diff
@@ -76,74 +77,96 @@ func (a *ArtifactDownloader) Download() error {
-			// See: http://golang.org/doc/effective_go.html#channels
-			artifact := artifact
-
-			p.Spawn(func() {
-				var err error
-				var path string = artifact.Path
-
-				// Convert windows paths to slashes, otherwise we get a literal
-				// download of "dir/dir/file" vs sub-directories on non-windows agents
-				if runtime.GOOS != `windows` {
-					path = strings.Replace(path, `\`, `/`, -1)
-				}
+	}
 
-				// Handle downloading from S3, GS, or RT
-				if strings.HasPrefix(artifact.UploadDestination, "s3://") {
-					err = NewS3Downloader(a.logger, S3DownloaderConfig{
-						Path:        path,
-						Bucket:      artifact.UploadDestination,
-						Destination: downloadDestination,
-						Retries:     5,
-						DebugHTTP:   a.conf.DebugHTTP,
-					}).Start()
-				} else if strings.HasPrefix(artifact.UploadDestination, "gs://") {
-					err = NewGSDownloader(a.logger, GSDownloaderConfig{
-						Path:        path,
-						Bucket:      artifact.UploadDestination,
-						Destination: downloadDestination,
-						Retries:     5,
-						DebugHTTP:   a.conf.DebugHTTP,
-					}).Start()
-				} else if strings.HasPrefix(artifact.UploadDestination, "rt://") {
-					err = NewArtifactoryDownloader(a.logger, ArtifactoryDownloaderConfig{
-						Path:        path,
-						Repository:  artifact.UploadDestination,
-						Destination: downloadDestination,
-						Retries:     5,
-						DebugHTTP:   a.conf.DebugHTTP,
-					}).Start()
-				} else {
-					err = NewDownload(a.logger, http.DefaultClient, DownloadConfig{
-						URL:         artifact.URL,
-						Path:        path,
-						Destination: downloadDestination,
-						Retries:     5,
-						DebugHTTP:   a.conf.DebugHTTP,
-					}).Start()
-				}
+	a.logger.Info("Found %d artifacts. Starting to download to: %s", artifactCount, downloadDestination)
+
+	p := pool.New(pool.MaxConcurrencyLimit)
+	errors := []error{}
+	s3Clients := map[string]*s3.S3{}
+
+	for _, artifact := range artifacts {
+		// Create new instance of the artifact for the goroutine
+		// See: http://golang.org/doc/effective_go.html#channels
+		artifact := artifact
+
+		p.Spawn(func() {
+			var err error
+			var path string = artifact.Path
+
+			// Convert windows paths to slashes, otherwise we get a literal
+			// download of "dir/dir/file" vs sub-directories on non-windows agents
+			if runtime.GOOS != `windows` {
+				path = strings.Replace(path, `\`, `/`, -1)
+			}
+
+			// Handle downloading from S3, GS, or RT
+			if strings.HasPrefix(artifact.UploadDestination, "s3://") {
+				// We want to have as few S3 clients as possible, as creating them is kind of an expensive operation
+				// But it's also theoretically possible that we'll have multiple artifacts with different S3 buckets, and each
+				// S3Client only applies to one bucket, so we need to store the S3 clients in a map, one for each bucket
+				bucketName, _ := ParseS3Destination(artifact.UploadDestination)
+				if _, has := s3Clients[bucketName]; !has {
+					p.Lock()
```

Comment:
I think this is open to a race condition, although perhaps a harmless one…

When multiple goroutines spawn concurrently for the same bucket, they'll each find is missing from the hash, and then take it in turns to wait/acquire the lock, instantiate a client, and place it in the `s3Clients` map.

When they pass `S3Client:    s3Clients[bucketName]` later on, it'll be unpredictable which instance will win.

I think the lock/unlock needs to go outside/around the `if _, has := s3Clients[bucketName]` bit?

Reply:
I would be tempted to do an entirely separate `for _, artifact := range artifacts {` prior to this one, just to prepare a `map[bucketstring]client` that the concurrent workers can treat as entirely read-only.

Reply:
I think you're right - part of me is like "oh no, it's O(2n) instead of O(n)!" but it pretty much doesn't matter at all compared to the amount of IO/etc we're doing. I'll do a cheeky lil pre-loop here.